kernel/cmdline.cc

```diff
@@ -23,11 +26,58 @@ cmdlineread(mdev*, char *dst, u32 off, u32 n)
   return cc;
 }
 
+// Returns true if param is found in cmdline, false otherwise.
+// If found, writes the value of the first occurence to dst.
+// Expects cmdline to be a space-delimeted list of <param>=<value> pairs.
+static bool
+getvalue(const char* param, char* dst)
+{
+  char parameq[CMDLINE_PARAM+1];
+  char *p, *end;
+
+  // find '<param>=' in cmdline
+  strcpy(parameq, param);
+  end = parameq + strlen(parameq);
+  *end++ = '=';
+  *end = 0;
+  p = strstr(cmdline, parameq);
+  if(p == NULL)
+    return false;
+
+  // copy <value> to dst
+  p += strlen(parameq);  // jump to after '='
+  while(*p != 0 && *p != ' ')
+    *dst++ = *p++;
+  *dst = 0;
+  return true;
+}
+
+// parse cmdline to populate global cmdline_params struct
+static void
+parsecmdline(void)
+{
+  char value[CMDLINE_VALUE];
+
+  if(getvalue("disable_pcid", value) && strcmp(value, "yes") == 0) {
+    cmdline_params.disable_pcid = true;
+    cprintf("cmdline: pcid disabled\n");
+  } else
+    cmdline_params.disable_pcid = false;
+
+  if(getvalue("keep_retpolines", value) && strcmp(value, "yes") == 0) {
+    cmdline_params.keep_retpolines = true;
+    cprintf("cmdline: retpolines not removed\n");
+  } else
+    cmdline_params.keep_retpolines = false;
+}
+
 void
 initcmdline(void)
 {
   if (VERBOSE)
     cprintf("cmdline: %s\n", cmdline);
```

Comment:
Have you verified that printing works this early in the boot process?

Reply:
Yep, at least when booting with qemu, the print statements appear in the console. I believe `inituart` sets up the console.